Implement the `kubeadm upgrade` command

luxas opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

Implements the kubeadm upgrades proposal: https://docs.google.com/document/d/1PRrC2tvB-p7sotIA5rnHy5WAOGdJJOIXPPv23hUFGrY/edit#

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes: https://github.com/kubernetes/kubeadm/issues/14

**Special notes for your reviewer**:

I'm gonna split out changes not directly related to the upgrade procedure into separate PRs as dependencies as we go.
**Now ready for review. Please look at `cmd/kubeadm/app/phases/upgrade` first and give feedback on that. The rest kind of follows.**

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Implemented `kubeadm upgrade plan` for checking whether you can upgrade your cluster to a newer version
Implemented `kubeadm upgrade apply` for upgrading your cluster from one version to an other
```

cc @fabriziopandini @kubernetes/sig-cluster-lifecycle-pr-reviews @craigtracey @mattmoyer

timothysc commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,34 @@
+
+func NewCmdApply() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "apply",
```
So I'm not sold on apply vs. : 
``` 
> kubectl upgrade 
or > kubectl upgrade --dry-run 
```

timothysc reviewed (changes requested):
Needs more input on the UX side before we go too far.

timothysc commented:
Per design doc discussion there will be an apply and plan and no dry-run. 

https://docs.google.com/document/d/1PRrC2tvB-p7sotIA5rnHy5WAOGdJJOIXPPv23hUFGrY/edit?disco=AAAAA_en9VI&ts=596e54c5

/cc @jbeda

luxas commented:
ping @timothysc @andrewrynhard @kubernetes/sig-cluster-lifecycle-feature-requests 

PTAL if you want/like. This is not the end state, it's work in progress.
There's a lot of functionality obviously, so I'm gonna split this up in separate PRs later on.
But first I'm gonna do everything in obe branch here, that people can easily checkout and test.

When I've written all necessary unit tests, added all necessary functionality for a MVP and cleaned up, this is ready for the real review.
However, it'd be great if you want to take a quick pass on a high level so I can catch things sooner rather than later.

So feel free to compile yourself and test :)
Only thing that's yet missing is the actual rolling upgrade code for the control plane.  Not sure if I should do it the hacky way now (I probably will, just to get things off the ground) or wait until we have "add first, then delete" update strategy for DaemonSets.

WDYT?

luxas commented:
This is now ready for review! Both `plan` and `apply` works, and most of the functionality is there.
I'll break out dependencies on refactoring and such things later into separate PRs.
**Please review `cmd/kubeadm/app/phases/upgrade` and the UX first**

UX:
```console
$ # --apiserver-bind-port is there to show that configuration is preserved with the upgrade. If it wouldn't, things would break in between when the API server would start listening on the default 6443, but this works.
$ # I've put kubeadm built from this PR in ./kubeadm
$ ./kubeadm init --kubernetes-version v1.7.1 --apiserver-bind-port 443
$ kubectl apply -f https://git.io/weave-kube-1.6
$ ./kubeadm alpha upgrade plan
[upgrade] Making sure the cluster is healthy:
[upgrade/health] Checking API Server health: Healthy
[upgrade/health] Checking Node health: All Nodes are healthy
[upgrade/health] Checking if control plane is Static Pod-hosted or Self-Hosted: Static Pod-hosted.
[upgrade/health] NOTE: kubeadm will upgrade your Static Pod-hosted control plane to a Self-Hosted one when upgrading if --feature-gates=SelfHosting=true is set (which is the default)
[upgrade/health] If you strictly want to continue using a Static Pod-hosted control plane, set --feature-gates=SelfHosting=true when running 'kubeadm upgrade apply'
[upgrade/health] Checking Static Pod manifests exists on disk: All required Static Pod manifests exist on disk
[upgrade] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster (you can get this with 'kubectl -n kube-system get cm kubeadm-config -oyaml')
[upgrade] Fetching available versions to upgrade to:
[upgrade/versions] Cluster version: v1.7.1
[upgrade/versions] kubeadm version: v1.8.0-alpha.2.789+11f48dc291fe93
[upgrade/versions] Latest stable version: v1.7.3
[upgrade/versions] Latest version in the v1.7 series: v1.7.3

Components that must be upgraded manually after you've upgraded the control plane with `kubeadm upgrade apply`:
COMPONENT   CURRENT      AVAILABLE
Kubelet     1 x v1.7.0   v1.7.3

Upgrade to the latest version in the v1.7 series:

COMPONENT            CURRENT   AVAILABLE
API Server           v1.7.1    v1.7.3
Controller Manager   v1.7.1    v1.7.3
Scheduler            v1.7.1    v1.7.3
Kube Proxy           v1.7.1    v1.7.3
Kube DNS             1.14.4    1.14.4

You can now apply the upgrade by executing the following command:

	kubeadm upgrade apply --version v1.7.3

_____________________________________________________________________

$ ./kubeadm alpha upgrade apply --version v1.7.3
[upgrade] Making sure the cluster is healthy:
[upgrade/health] Checking API Server health: Healthy
[upgrade/health] Checking Node health: All Nodes are healthy
[upgrade/health] Checking if control plane is Static Pod-hosted or Self-Hosted: Static Pod-hosted.
[upgrade/health] NOTE: kubeadm will upgrade your Static Pod-hosted control plane to a Self-Hosted one when upgrading if --feature-gates=SelfHosting=true is set (which is the default)
[upgrade/health] If you strictly want to continue using a Static Pod-hosted control plane, set --feature-gates=SelfHosting=true when running 'kubeadm upgrade apply'
[upgrade/health] Checking Static Pod manifests exists on disk: All required Static Pod manifests exist on disk
[upgrade] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster (you can get this with 'kubectl -n kube-system get cm kubeadm-config -oyaml')
[upgrade/version] You have chosen to upgrade to version "v1.7.3"
[upgrade/versions] Cluster version: v1.7.1
[upgrade/versions] kubeadm version: v1.8.0-alpha.2.789+11f48dc291fe93
[upgrade/confirm] Are you sure you want to proceed with the upgrade? [y/N]: Y
[upgrade/prepull] Will prepull images for components [kube-apiserver kube-controller-manager kube-scheduler]
[upgrade/prepull] Prepulling image for component kube-scheduler.
[upgrade/prepull] Prepulling image for component kube-apiserver.
[upgrade/prepull] Prepulling image for component kube-controller-manager.
[upgrade/prepull] Prepulled image for component kube-scheduler.
[upgrade/prepull] Prepulled image for component kube-apiserver.
[upgrade/prepull] Prepulled image for component kube-controller-manager.
[upgrade/prepull] Successfully prepulled the images for all the control plane components
[upgrade/apply] Upgrading your Static Pod-hosted control plane to version "v1.7.3"...
[upgrade/staticpods] Wrote upgraded Static Pod manifests to "/tmp/kubeadm-upgrade830923296"
[upgrade/staticpods] Moved upgraded manifest to "/etc/kubernetes/manifests/kube-apiserver.yaml" and backuped old manifest to "/tmp/kubeadm-upgrade830923296/old-manifests/kube-apiserver.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[apiclient] Found 1 Pods for label selector component=kube-apiserver
[upgrade/staticpods] Component "kube-apiserver" upgraded successfully!
[upgrade/staticpods] Moved upgraded manifest to "/etc/kubernetes/manifests/kube-controller-manager.yaml" and backuped old manifest to "/tmp/kubeadm-upgrade830923296/old-manifests/kube-controller-manager.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[apiclient] Found 1 Pods for label selector component=kube-controller-manager
[upgrade/staticpods] Component "kube-controller-manager" upgraded successfully!
[upgrade/staticpods] Moved upgraded manifest to "/etc/kubernetes/manifests/kube-scheduler.yaml" and backuped old manifest to "/tmp/kubeadm-upgrade830923296/old-manifests/kube-scheduler.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[apiclient] Found 1 Pods for label selector component=kube-scheduler
[apiclient] Found 0 Pods for label selector component=kube-scheduler
[apiclient] Found 1 Pods for label selector component=kube-scheduler
[upgrade/staticpods] Component "kube-scheduler" upgraded successfully!
[self-hosted] Created TLS secret "ca" from ca.crt and ca.key
[self-hosted] Created TLS secret "apiserver" from apiserver.crt and apiserver.key
[self-hosted] Created TLS secret "apiserver-kubelet-client" from apiserver-kubelet-client.crt and apiserver-kubelet-client.key
[self-hosted] Created TLS secret "sa" from sa.pub and sa.key
[self-hosted] Created TLS secret "front-proxy-ca" from front-proxy-ca.crt and front-proxy-ca.key
[self-hosted] Created TLS secret "front-proxy-client" from front-proxy-client.crt and front-proxy-client.key
[self-hosted] Created secret "scheduler.conf"
[self-hosted] Created secret "controller-manager.conf"
[apiclient] Found 0 Pods for label selector k8s-app=self-hosted-kube-apiserver
[apiclient] Found 1 Pods for label selector k8s-app=self-hosted-kube-apiserver
[self-hosted] self-hosted kube-apiserver ready after 15.295571 seconds
[apiclient] Found 0 Pods for label selector k8s-app=self-hosted-kube-controller-manager
[apiclient] Found 1 Pods for label selector k8s-app=self-hosted-kube-controller-manager
[self-hosted] self-hosted kube-controller-manager ready after 15.804586 seconds
[apiclient] Found 0 Pods for label selector k8s-app=self-hosted-kube-scheduler
[apiclient] Found 1 Pods for label selector k8s-app=self-hosted-kube-scheduler
[self-hosted] self-hosted kube-scheduler ready after 18.958074 seconds
[apiconfig] Created RBAC rules
[addons] Applied essential addon: kube-proxy
[addons] Applied essential addon: kube-dns
```

PTAL and heads up @kubernetes/sig-cluster-lifecycle-pr-reviews @fabriziopandini @andrewrynhard @aaronlevy @ericchiang @mattmoyer @jamiehannaford @pipejakob @justinsb @jnardiello @craigtracey @kad @nckturner @idvoretskyi @calebamiles @jdumars

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -0,0 +1,252 @@
+	// This guess will be corrected once we know if there is a patch version available
+	canDoMinorUpgrade := clusterVersion.LessThan(stableVersion)
+
+	// A patch version doesn't exist if the cluster version is higher than the stable version
```
higher than or equal to the current stable version

luxas replied:
Fixed

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -0,0 +1,252 @@
+	canDoMinorUpgrade := clusterVersion.LessThan(stableVersion)
+
+	// A patch version doesn't exist if the cluster version is higher than the stable version
+	// In the case an user is trying to upgrade from, let's say, v1.8.0-beta.2 to v1.8.0-rc.1 (given we support such upgrades experimentally)
```
in the case that a user

luxas replied:
Fixed

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/versiongetter.go:
```diff
@@ -0,0 +1,139 @@
+		fmt.Fprintf(g.w, "[upgrade/versions] FATAL: Couldn't parse cluster version: %v\n", err)
+		return "", nil, err
+	}
+	return clusterVersionInfo.String(), clusterVersion, nil
```
do we need to return an extra string arg? Can't we call `clusterVersion.String()`?

luxas replied:
Unfortunately they return different results at this time :(
`clusterVersionInfo.String()` etc contains the `v` prefix and possibly more information like build metadata, etc., which is stripped away when converting it to `*version.Version`. One could argue to improve the `version` package though.

WDYT?

jamiehannaford replied:
I think for now it's fine to keep as is, but maybe we can file an issue with our use case and also add a `TODO` here.

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -0,0 +1,252 @@
+		description := fmt.Sprintf("version in the v%s series", currentBranch)
+
+		// Get and output the latest patch version for the cluster branch
+		patchVersionStr, patchVersion, err := versionGetterImpl.GetLatestKubeVersion(versionLabel, description)
```
to continue the above comment, imagine the current cluster version is `v1.8.0-beta.2`, would `clusterVersionStr` be `v1.8.0`? because if not `currentBranch` would be `ta.2` (the last 3 chars)

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -0,0 +1,252 @@
+		}
+
+		// Check if a minor version upgrade is possible when a patch release exists
+		// It's only possible if the latest stable version is higher than the current patch version
```
can we change this to: `It's only possible if the latest patch version is higher ...`

luxas replied:
Fixed

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -0,0 +1,252 @@
+	// A patch version doesn't exist if the cluster version is higher than the stable version
+	// In the case an user is trying to upgrade from, let's say, v1.8.0-beta.2 to v1.8.0-rc.1 (given we support such upgrades experimentally)
+	// a stable-1.8 branch doesn't exist yet. Hence this check.
+	if patchVersionBranchExists(clusterVersion, stableVersion) {
```
if they want to upgrade from v1.8.0-beta.2 to v1.8.0-rc.1 _after_ v1.8 is released, I assume the current stable version would be `v1.8`, meaning that this conditional would pass, right?

jamiehannaford reviewed (changes requested):
First round of feedback done (mostly questions and small nit comments). I'll take another look tomorrow.

luxas commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -96,6 +97,24 @@ const (
+	DefaultEtcdVersion = "3.0.17"
+
+	// MasterConfigurationConfigMap specifies in what ConfigMap in the kube-system namespace the `kubeadm init` configuration should be stored
+	MasterConfigurationConfigMap = "kubeadm-config"
```
This merged in an other PR: https://github.com/kubernetes/kubernetes/pull/50320.
I kind of agree, but we can use different keys for different types of configuration, so I don't think it's a huge problem.
(Init config under the `MasterConfiguration` key, etc.)

jamiehannaford replied:
Okay, fair enough

luxas commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -118,11 +137,29 @@ var (
-func BuildStaticManifestFilepath(componentName string) string {
-	return filepath.Join(KubernetesDir, ManifestsSubDirName, componentName+".yaml")
+// GetStaticPodDirectory returns the location on the disk where the Static Pod should be present
+func GetStaticPodDirectory() string {
```
What would you prefer? This merged in https://github.com/kubernetes/kubernetes/pull/50173
Feel free to send a PR if you have opinions ;)

jamiehannaford replied:
Just the function without the `Get` prefix. Not a big deal ;)

luxas commented on cmd/kubeadm/app/phases/saveconfig/saveconfig.go:
```diff
@@ -0,0 +1,60 @@
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// SaveConfiguration saves the MasterConfiguration used for later reference (when upgrading for instance)
```
I think this was changed to something better in https://github.com/kubernetes/kubernetes/pull/50320

luxas commented on cmd/kubeadm/app/util/apiclient.go:
```diff
@@ -59,22 +63,27 @@ func WaitForAPI(client *clientset.Clientset) {
+		pods, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(listOpts)
 		if err != nil {
-			fmt.Printf("[apiclient] Error getting Pods with label selector %q [%v]\n", labelKeyValPair, err)
+			fmt.Fprintf(out, "[apiclient] Error getting Pods with label selector %q [%v]\n", labelKeyValPair, err)
```
I think so, yeah. Do you have an other suggestion?

jamiehannaford replied:
Actually I think it's fine. For some reason I associated it with the apiserver, but I guess we're just indicating it's a client-related process.

luxas commented on cmd/kubeadm/app/phases/saveconfig/saveconfig.go:
```diff
@@ -0,0 +1,60 @@
+		},
+	}
+
+	if _, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Create(&cm); err != nil {
```
Kind of agree, but at the same time I'm a bit hesitant to add more moving parts into the system.
This config isn't the thing we have to guard the most.

jamiehannaford replied:
Is or isn't? I thought this ConfigMap was pretty important. I wouldn't say a namespace is a moving part, just an organizational detail, but I take your point. Maybe we can add one if further kubeadm-specific resources are added in the future.

luxas replied:
Yeah, it always is possible to change later in the cycle or in a future release.
Would you mind open an issue for this though so we can track it and get wider visibility?

luxas commented on cmd/kubeadm/app/util/apiclient.go:
```diff
@@ -59,22 +63,27 @@ func WaitForAPI(client *clientset.Clientset) {
 
-		if len(apiPods.Items) == 0 {
+		if lastKnownPodNumber != len(pods.Items) {
+			fmt.Fprintf(out, "[apiclient] Found %d Pods for label selector %s\n", len(pods.Items), labelKeyValPair)
```
I've now changed it to only report on a change; I now think they make sense.
(They didn't before indeed)

luxas commented on cmd/kubeadm/app/phases/upgrade/versiongetter.go:
```diff
@@ -0,0 +1,139 @@
+	versionStr, err := kubeadmutil.KubernetesReleaseVersion(ciVersionLabel)
+	if err != nil {
+		if description != "" {
+			fmt.Fprintln(g.w, "<notfound>")
```
yes. Since the previous print was intentionally a `Fprintf` without an `\n`, this is needed to inform about the error and make a new line in order for us to be able to print a normal line on the next print.

jamiehannaford replied:
How about just an empty `fmt.Fprintln(g.w, "")`. I'm just thinking the `<notfound>` message seems a bit incongruous with our formatting

luxas commented on cmd/kubeadm/app/phases/upgrade/health.go:
```diff
@@ -0,0 +1,221 @@
+	} else {
+		fmt.Println("Static Pod-hosted.")
+		fmt.Println("[upgrade/health] NOTE: kubeadm will upgrade your Static Pod-hosted control plane to a Self-Hosted one when upgrading if --feature-gates=SelfHosting=true is set (which is the default)")
+		fmt.Println("[upgrade/health] If you strictly want to continue using a Static Pod-hosted control plane, set --feature-gates=SelfHosting=true when running 'kubeadm upgrade apply'")
```
Totally missed that's not the case. I'm aiming to expose the feature flags in the CLI on both `init` and `upgrade`. Good catch

jamiehannaford commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -0,0 +1,174 @@
+	for _, upgrade := range upgrades {
+
+		if upgrade.CanUpgradeKubelets() {
+			fmt.Println("Components that must be upgraded manually after you've upgraded the control plane with `kubeadm upgrade apply`:")
```
should this be `Fprintln`?

luxas replied:
no; the writer w is only for tab output

jamiehannaford commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -0,0 +1,174 @@
+		fmt.Println("")
+
+		if upgrade.Before.KubeadmVersion != upgrade.After.KubeadmVersion {
+			fmt.Printf("Note: Before you do can perform this upgrade, you have to update kubeadm to %s\n", upgrade.After.KubeadmVersion)
```
Can we instruct users how to do this? Or maybe link off to a doc?

jamiehannaford commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,172 @@
+		}
+	}
+
+	if !flags.flagYes && !flags.dryRun {
```
so imagine I pass in `--dry-run` and leave out `--yes`, where is the logic to perform the dry run?

timothysc replied:
I'm still confused by (yes).

jamiehannaford commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,172 @@
+		},
+	}
+
+	cmd.Flags().BoolVar(&flags.flagYes, "yes", flags.flagYes, "Actually perform the selected upgrade")
```
Maybe change this to `Perform the upgrade and do not prompt for confirmation`

luxas replied:
fixed

jamiehannaford commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,172 @@
+	if versionSkewErrs != nil {
+
+		if len(versionSkewErrs.Mandatory) > 0 {
+			fmt.Printf("[upgrade/version] FATAL: The --version argument is invalid: %v.\n[upgrade/version] Can be bypassed if you pass the --force flag\n", versionSkewErrs.Mandatory)
```
Is this correct? I thought mandatory errors could not be bypassed with `--force`?

luxas replied:
that was a typo. fixed

jamiehannaford commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,172 @@
+			return fmt.Errorf("The --version argument is invalid")
+		}
+
+		if len(versionSkewErrs.Skippable) > 0 {
```
Can we actually format and return the version skews so the user knows why the provided version is invalid? Right now we're asking them to figure out the mystery by themselves and looking at the logic in `EnforceVersionPolicies` that might be a difficult task 🔎

jamiehannaford commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,172 @@
+			if !flags.force {
+				return fmt.Errorf("The --version argument is invalid")
+			}
+			fmt.Println("[upgrade/version] Continuing since the --force flag is set")
```
Maybe also mention how many skippable errors were found. `Found %d potential compatibility errors but skipping since ...`

luxas replied:
fixed, thanks

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/prepull.go:
```diff
@@ -0,0 +1,177 @@
+	fmt.Printf("[upgrade/prepull] Prepulled image for component %s.\n", component)
+}
+
+// PrepullImagesInParallell creates DaemonSets synchronously but waits in parallell for the images to pull
```
parallel

luxas replied:
fixed

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/prepull.go:
```diff
@@ -0,0 +1,177 @@
+						{
+							Name:    component,
+							Image:   images.GetCoreImage(component, cfg.ImageRepository, cfg.KubernetesVersion, cfg.UnifiedControlPlaneImage),
+							Command: []string{"sleep", "3600"},
```
Initially I was wondering whether 60s would be enough for user's on a really slow box (hyperkube images can be over 600MB) but I guess the container command doesn't affect the kubelet's image pull, right?

luxas replied:
No this doesn't affect that. This is just a dummy command

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/prepull.go:
```diff
@@ -0,0 +1,177 @@
+	deleteOptions := &metav1.DeleteOptions{
+		PropagationPolicy: &foregroundDelete,
+	}
+	if err := d.client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Delete(dsName, deleteOptions); err != nil {
```
should we return an error here? In other words, should a failed cleanup halt execution?

jamiehannaford commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,172 @@
+		fmt.Printf("[upgrade/apply] Upgrading your Self-Hosted control plane to version %q...\n", flags.newK8sVersionStr)
+
+		// Upgrade a self-hosted cluster
+		return fmt.Errorf("not implemented")
```
maybe we should fail earlier than this for self-hosted clusters? I'd rather know straight away that's it's not yet implemented than after a few minutes of optimistic expectation

jamiehannaford replied:
BTW do we expect to have this working for 1.8?

jamiehannaford commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,172 @@
+		}
+
+		// Upgrade the static pod hosted control plane to a self-hosted one
+		// TODO: Do this only in case the self-hosting feature gate is enabled
```
just a reminder that we can do that after a rebase

luxas replied:
now fixed; we won't upgrade a static pod cluster to a self-hosted one in v1.8, so I could just remove this

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -0,0 +1,127 @@
+// UpgradeStaticPodControlPlane upgrades a static pod-hosted control plane
+func UpgradeStaticPodControlPlane(client clientset.Interface, cfg *kubeadmapi.MasterConfiguration, newK8sVersion *version.Version) error {
+
+	// This string-string map stores the component name and backup filepath.
```
filepath (if a rollback is needed).

luxas replied:
fixed

jamiehannaford commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -118,11 +137,29 @@ var (
+}
+
+// GetStaticPodFilepath returns the location on the disk where the Static Pod should be present
+func GetStaticPodFilepath(componentName, manifestsDir string) string {
```
nit: isn't it convention to put the dir arg before the filename?

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -0,0 +1,127 @@
+			return rollbackOldManifests(recoverManifests, err)
+		}
+
+		fmt.Printf("[upgrade/staticpods] Moved upgraded manifest to %q and backuped old manifest to %q\n", currentManifestPath, backupManifestPath)
```
backed up 😛

luxas replied:
fixed

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -0,0 +1,127 @@
+		// Wait for the mirror Pod hash to change; otherwise we'll run into race conditions here when the kubelet hasn't had time to
+		// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+		// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+		// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
```
👍  great comment. We should always aspire to document algorithms like this

jamiehannaford commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -0,0 +1,127 @@
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
```
is it worth waiting until the recovered pod is healthy? because if we put it back and it doesn't come back online, the user has _another_ thing to worry about. A failed upgrade is one thing, but a failed upgrade which brings down your cluster is another

luxas replied:
I don't think there is any upgrade procedure that is totally 100% safe to take, as the code is written by humans ;)
Anyway, rolling back is on a best-effort; do you mean there would be anything more we could do?
What would we do if the old manifest _doesn't_ come up?

pipejakob replied:
I definitely haven't been involved enough in the design or these discussions, but have a long personal history of doing large migrations that were carefully designed to be safe. For instance, if you're doing a data migration from store A to store B, a safe plan could be something like the following:

1) Keep all read operations using A, but change writes so they propagate to both A and B simultaneously.
2) Backfill any data that exists in A and is missing in B into B.
3) Run a consistency checker that is constantly probing to make sure there aren't discrepancies between data in A and B.
4) Change read operations to read from A and B simultaneously. Always return the value from A, but create a silent report if the values of A and B didn't match.
5) When enough time has passed and you're certain that B is healthy and up to date, change reads to only read from B.
6) Change writes to only write to B.
7) Completely retire A.

It might seem overly complex, especially since each step is a different launch that is incrementally getting you closer to the full migration, but it has the nice property that at any point, if something goes wrong, you can always stop the migration or roll back to previous steps without any data loss. There is no single "flip over to B" step that has a window of uncertainty or potential data loss if things don't go well. You never move onto the next step until the previous step has proved successful, and you can stop safely at any time, and move backwards in the plan at any time. Contrast this to the fast-but-easy plan of taking scheduled downtime to flip everything over to point to B all at once, and crossing your fingers that it all works when you bring the system back online.

It would be great if our upgrade plan had similar properties, so that kubeadm or any random component could crash at any time during the upgrade, and we would never leave the cluster in an unusable state where human intervention would be required to fix it. But, I don't know how feasible that is with the mechanisms we have today, and don't have specific suggestions on how to get there now, so this is just a general grumbling that I hope we can figure out how to do better.

If I have time to offer concrete suggestions for how to do these steps differently, I'll let you know. :-)

jamiehannaford reviewed (changes requested):
Some more feedback. Is it worth having e2e tests for this? I guess that might take a lot of work, so maybe we should just add it as an issue for now.

luxas commented on cmd/kubeadm/app/cmd/upgrade/upgrade.go:
```diff
@@ -0,0 +1,72 @@
+
+	cmd.PersistentFlags().StringVar(&flags.kubeConfigPath, "kubeconfig", flags.kubeConfigPath, "The KubeConfig file to use for talking to the cluster")
+	cmd.PersistentFlags().StringVar(&flags.cfgPath, "config", flags.cfgPath, "Path to kubeadm config file (WARNING: Usage of a configuration file is experimental)")
+	cmd.PersistentFlags().BoolVar(&flags.allowExperimentalUpgrades, "allow-experimental-upgrades", flags.allowExperimentalUpgrades, "Show unstable versions as an upgrade alternative and allow upgrading to an unstable version")
```
Fixed

luxas commented on cmd/kubeadm/app/cmd/upgrade/upgrade.go:
```diff
@@ -0,0 +1,72 @@
+	cmd.PersistentFlags().StringVar(&flags.kubeConfigPath, "kubeconfig", flags.kubeConfigPath, "The KubeConfig file to use for talking to the cluster")
+	cmd.PersistentFlags().StringVar(&flags.cfgPath, "config", flags.cfgPath, "Path to kubeadm config file (WARNING: Usage of a configuration file is experimental)")
+	cmd.PersistentFlags().BoolVar(&flags.allowExperimentalUpgrades, "allow-experimental-upgrades", flags.allowExperimentalUpgrades, "Show unstable versions as an upgrade alternative and allow upgrading to an unstable version")
+	cmd.PersistentFlags().BoolVar(&flags.allowRCUpgrades, "allow-release-candidate-upgrades", flags.allowRCUpgrades, "Show release candidate versions as an upgrade alternative and allow upgrading to a release candidate version")
```
Fixed

luxas commented on cmd/kubeadm/app/cmd/upgrade/upgrade.go:
```diff
@@ -0,0 +1,72 @@
+	cmd.PersistentFlags().StringVar(&flags.cfgPath, "config", flags.cfgPath, "Path to kubeadm config file (WARNING: Usage of a configuration file is experimental)")
+	cmd.PersistentFlags().BoolVar(&flags.allowExperimentalUpgrades, "allow-experimental-upgrades", flags.allowExperimentalUpgrades, "Show unstable versions as an upgrade alternative and allow upgrading to an unstable version")
+	cmd.PersistentFlags().BoolVar(&flags.allowRCUpgrades, "allow-release-candidate-upgrades", flags.allowRCUpgrades, "Show release candidate versions as an upgrade alternative and allow upgrading to a release candidate version")
+	cmd.PersistentFlags().BoolVar(&flags.printConfig, "print-config", flags.printConfig, "Whether the configuration file that will be used in the upgrade should be printed or not")
```
We/I might add a dedicated `kubeadm config` command as part of this...

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,172 @@
+
+	"github.com/spf13/cobra"
+
+	//clientset "k8s.io/client-go/kubernetes"
```
Fixed

luxas commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -0,0 +1,174 @@
+func NewCmdPlan(parentFlags *cmdUpgradeFlags) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "plan",
+		Short: "Check which versions are available to upgrade to",
```
Fixed

luxas commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -0,0 +1,174 @@
+
+	// Run healthchecks against the cluster
+	if !upgrade.VerifyClusterHealth(client) {
+		return nil, nil, fmt.Errorf("the cluster is not in an upgradeable state")
```
Fixed

luxas commented on cmd/kubeadm/app/phases/upgrade/health.go:
```diff
@@ -0,0 +1,221 @@
+	}
+
+	// If the control plane is self-hosted, we should run more health checks
+	fmt.Printf("[upgrade/health] Checking if control plane is Static Pod-hosted or Self-Hosted: ")
```
Outcommented for now; can revisit later. I think it makes sense to leave this detail out

luxas commented on cmd/kubeadm/app/phases/upgrade/health.go:
```diff
@@ -0,0 +1,221 @@
+func nodesHealthy(client clientset.Interface) error {
+	nodes, err := client.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("couldn't list all nodes in cluster")
```
Fixed

luxas commented on cmd/kubeadm/app/phases/upgrade/configuration.go:
```diff
@@ -0,0 +1,95 @@
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// FetchConfiguration fetches configuration required for upgrading your cluster from a file which has precendence or a ConfigMap in the cluster
```
Fixed

luxas commented:
Thanks for the review @jamiehannaford :heart: !
There are ~10-15-ish open comments that I'll resolve/take a look at later when I have time.

timothysc reviewed (changes requested):
This is a pretty big pill to swallow in a single thunk.   
@lukemarsden you might have comments on flow and feel here.

timothysc commented:
Perhaps you could start with just apply then add the options separate.

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// RunApply takes care of the actual upgrade functionality
+func RunApply(flags *applyFlags) error {
```
fixed

luxas commented:
I've now rebased this upon all the sub-PRs I've broken out if this and got merged. More to come, but here is a status update on the UX. I'm running all this as `root` and kubeadm compiled from this PR is available at `./kubeadm`. You can try all this yourself if you download the PR and test:

```console
root@THEGOPHER:~/luxas/kubernetes# ./kubeadm init --apiserver-bind-port 443 --kubernetes-version v1.7.2
[kubeadm] WARNING: kubeadm is in beta, please do not use it for production clusters.
[init] Using Kubernetes version: v1.7.2
[init] Using Authorization mode: [Node RBAC]
[preflight] Running pre-flight checks
[preflight] WARNING: docker service is not enabled, please run 'systemctl enable docker.service'
[preflight] Starting the kubelet service
[kubeadm] WARNING: starting in 1.8, tokens expire after 24 hours by default (if you require a non-expiring token use --token-ttl 0)
[certificates] Generated CA certificate and key.
[certificates] Generated API server certificate and key.
[certificates] API server serving cert is signed for DNS names [thegopher kubernetes kubernetes.default kubernetes.default.svc kubernetes.default.svc.cluster.local] and IPs [10.96.0.1 192.168.1.130]
[certificates] Generated API server kubelet client certificate and key.
[certificates] Generated service account key and public key.
[certificates] Generated front-proxy CA certificate and key.
[certificates] Generated front-proxy client certificate and key.
[certificates] Valid certificates and keys now exist in "/etc/kubernetes/pki"
[kubeconfig] Wrote KubeConfig file to disk: "admin.conf"
[kubeconfig] Wrote KubeConfig file to disk: "kubelet.conf"
[kubeconfig] Wrote KubeConfig file to disk: "controller-manager.conf"
[kubeconfig] Wrote KubeConfig file to disk: "scheduler.conf"
[init] Waiting for the kubelet to boot up the control plane as Static Pods from directory "/etc/kubernetes/manifests"
[apiclient] All control plane components are healthy after 49.501170 seconds
[markmaster] Master thegopher tainted and labelled with key/value: node-role.kubernetes.io/master=""
[token] Using token: cca24e.b856730fd78747e5
[bootstraptoken] Configured RBAC rules to allow Node Bootstrap tokens to post CSRs in order for nodes to get long term certificate credentials
[bootstraptoken] Configured RBAC rules to allow the csrapprover controller automatically approve CSRs from a Node Bootstrap Token
[bootstraptoken] Creating the "cluster-info" ConfigMap in the "kube-public" namespace
[uploadconfig] Storing the configuration used in ConfigMap "kubeadm-config" in the "kube-system" Namespace
[addons] Applied essential addon: kube-dns
[addons] Applied essential addon: kube-proxy

Your Kubernetes master has initialized successfully!

To start using your cluster, you need to run (as a regular user):

  mkdir -p $HOME/.kube
  sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
  sudo chown $(id -u):$(id -g) $HOME/.kube/config

You should now deploy a pod network to the cluster.
Run "kubectl apply -f [podnetwork].yaml" with one of the options listed at:
  http://kubernetes.io/docs/admin/addons/

You can now join any number of machines by running the following on each node
as root:

  kubeadm join --token cca24e.b856730fd78747e5 192.168.1.130:443 --discovery-token-ca-cert-hash sha256:c4b358699cc4fdc8a178c0b723bd76059360b1ae7a89259bf6f152f08072577b

root@THEGOPHER:~/luxas/kubernetes# export KUBECONFIG=/etc/kubernetes/admin.conf 
root@THEGOPHER:~/luxas/kubernetes# kubectl apply -f https://git.io/weave-kube-1.6
serviceaccount "weave-net" created
clusterrole "weave-net" created
clusterrolebinding "weave-net" created
daemonset "weave-net" created
root@THEGOPHER:~/luxas/kubernetes# ./kubeadm upgrade plan
[upgrade] Making sure the cluster is healthy:
[upgrade/health] Checking API Server health: Healthy
[upgrade/health] Checking Node health: All Nodes are healthy
[upgrade/health] Checking Static Pod manifests exists on disk: All required Static Pod manifests exist on disk
[upgrade] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[upgrade] Fetching available versions to upgrade to:
[upgrade/versions] Cluster version: v1.7.2
[upgrade/versions] kubeadm version: v1.8.0-alpha.2.2002+cab49bed7e9448
[upgrade/versions] Latest stable version: v1.7.3
[upgrade/versions] Latest version in the v1.7 series: v1.7.3

Upgrade to the latest version in the v1.7 series:

COMPONENT            CURRENT   AVAILABLE
API Server           v1.7.2    v1.7.3
Controller Manager   v1.7.2    v1.7.3
Scheduler            v1.7.2    v1.7.3
Kube Proxy           v1.7.2    v1.7.3
Kube DNS             1.14.4    1.14.4

You can now apply the upgrade by executing the following command:

	kubeadm upgrade apply --version v1.7.3

_____________________________________________________________________

root@THEGOPHER:~/luxas/kubernetes# ./kubeadm upgrade apply --version v1.7.3
[upgrade] Making sure the cluster is healthy:
[upgrade/health] Checking API Server health: Healthy
[upgrade/health] Checking Node health: All Nodes are healthy
[upgrade/health] Checking Static Pod manifests exists on disk: All required Static Pod manifests exist on disk
[upgrade] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[upgrade/version] You have chosen to upgrade to version "v1.7.3"
[upgrade/versions] Cluster version: v1.7.2
[upgrade/versions] kubeadm version: v1.8.0-alpha.2.2002+cab49bed7e9448
[upgrade/confirm] Are you sure you want to proceed with the upgrade? [y/N]: Y
[upgrade/prepull] Will prepull images for components [kube-apiserver kube-controller-manager kube-scheduler]
[upgrade/prepull] Prepulling image for component kube-scheduler.
[upgrade/prepull] Prepulling image for component kube-apiserver.
[upgrade/prepull] Prepulling image for component kube-controller-manager.
[upgrade/prepull] Prepulled image for component kube-scheduler.
[upgrade/prepull] Prepulled image for component kube-controller-manager.
[upgrade/prepull] Prepulled image for component kube-apiserver.
[upgrade/prepull] Successfully prepulled the images for all the control plane components
[upgrade/apply] Upgrading your Static Pod-hosted control plane to version "v1.7.3"...
[upgrade/staticpods] Wrote upgraded Static Pod manifests to "/tmp/kubeadm-upgrade213583318"
[upgrade/staticpods] Moved upgraded manifest to "/etc/kubernetes/manifests/kube-apiserver.yaml" and backed up old manifest to "/tmp/kubeadm-upgrade213583318/old-manifests/kube-apiserver.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[apiclient] Found 1 Pods for label selector component=kube-apiserver
[upgrade/staticpods] Component "kube-apiserver" upgraded successfully!
[upgrade/staticpods] Moved upgraded manifest to "/etc/kubernetes/manifests/kube-controller-manager.yaml" and backed up old manifest to "/tmp/kubeadm-upgrade213583318/old-manifests/kube-controller-manager.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[apiclient] Found 1 Pods for label selector component=kube-controller-manager
[apiclient] Found 0 Pods for label selector component=kube-controller-manager
[apiclient] Found 1 Pods for label selector component=kube-controller-manager
[upgrade/staticpods] Component "kube-controller-manager" upgraded successfully!
[upgrade/staticpods] Moved upgraded manifest to "/etc/kubernetes/manifests/kube-scheduler.yaml" and backed up old manifest to "/tmp/kubeadm-upgrade213583318/old-manifests/kube-scheduler.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[apiclient] Found 1 Pods for label selector component=kube-scheduler
[upgrade/staticpods] Component "kube-scheduler" upgraded successfully!
[bootstraptoken] Configured RBAC rules to allow Node Bootstrap tokens to post CSRs in order for nodes to get long term certificate credentials
[bootstraptoken] Configured RBAC rules to allow the csrapprover controller automatically approve CSRs from a Node Bootstrap Token
[addons] Applied essential addon: kube-dns
[addons] Applied essential addon: kube-proxy
[uploadconfig] Storing the configuration used in ConfigMap "kubeadm-config" in the "kube-system" Namespace

[upgrade/successful] SUCCESS! Your cluster was upgraded to "v1.7.3". Enjoy!

[upgrade/kubelet] Now that your control plane is upgraded, please proceed with upgrading your kubelets in turn.
root@THEGOPHER:~/luxas/kubernetes# 
```

@timothysc The next step is maybe to create a PR with stubs only in `phases/upgrade`, but with all the CLI logic ready. Sounds good? Then we'll tackle the business logic after that...

pipejakob commented:
> This is a pretty big pill to swallow in a single thunk.

Seconded. I've been trying to make my way through it, but it's a 2,800 line diff, and there have already been many comments so far to catch up on. I really like the idea of starting with the UX stubs as a first PR and incrementally adding logic in follow-up PRs. All of the policy code seems like a good candidate for an isolated PR, and a separate one just for the prepulling functionality, another for static pod upgrading, etc. I would be more than happy to review them one at a time when you've broken it up.

pipejakob commented on cmd/kubeadm/app/phases/upgrade/versiongetter.go:
```diff
@@ -0,0 +1,139 @@
+// versionGetter defines an interface for fetching different versions.
+// Easy to implement a fake variant of this interface for unit testing
+type versionGetter interface {
+	GetClusterVersion() (string, *versionutil.Version, error)
```
I know that we're pretty bad about this in the existing codebase, but I like the advice here of omitting the redundant "Get" prefix:

https://golang.org/doc/effective_go.html#Getters

pipejakob commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -0,0 +1,127 @@
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
```
There definitely seem to be a lot of steps that would leave the cluster in an unusable state if we accidentally panic/crash/lose power at the right time. I need to re-review the original proposal, but it would be a really nice property if every step was completely safe to take, without any of these danger windows that could leave you with an unusable control plane.

luxas commented:
@pipejakob See https://github.com/kubernetes/kubernetes/pull/50872

luxas commented:
@timothysc @pipejakob Rebased this. Last commit (implementation only) is down on ~2300 LoC (at least half of it is unit testing)

The five other commits are from other PRs I have up

timothysc commented:
Could you rebase this once the other PRs have merged.

luxas commented:
Moving a conversation from hidden comments on how to make the rollback of static pods safer:
```
@jamiehannaford:
is it worth waiting until the recovered pod is healthy? because if we put it back and it doesn't come back online, the user has another thing to worry about. A failed upgrade is one thing, but a failed upgrade which brings down your cluster is another

@pipejakob:
There definitely seem to be a lot of steps that would leave the cluster in an unusable state if we accidentally panic/crash/lose power at the right time. I need to re-review the original proposal, but it would be a really nice property if every step was completely safe to take, without any of these danger windows that could leave you with an unusable control plane.
@luxas:
I don't think there is any upgrade procedure that is totally 100% safe to take, as the code is written by humans ;)
Anyway, rolling back is on a best-effort; do you mean there would be anything more we could do?
What would we do if the old manifest doesn't come up?

@pipejakob:
I definitely haven't been involved enough in the design or these discussions, but have a long personal history of doing large migrations that were carefully designed to be safe. For instance, if you're doing a data migration from store A to store B, a safe plan could be something like the following:

Keep all read operations using A, but change writes so they propagate to both A and B simultaneously.
Backfill any data that exists in A and is missing in B into B.
Run a consistency checker that is constantly probing to make sure there aren't discrepancies between data in A and B.
Change read operations to read from A and B simultaneously. Always return the value from A, but create a silent report if the values of A and B didn't match.
When enough time has passed and you're certain that B is healthy and up to date, change reads to only read from B.
Change writes to only write to B.
Completely retire A.
It might seem overly complex, especially since each step is a different launch that is incrementally getting you closer to the full migration, but it has the nice property that at any point, if something goes wrong, you can always stop the migration or roll back to previous steps without any data loss. There is no single "flip over to B" step that has a window of uncertainty or potential data loss if things don't go well. You never move onto the next step until the previous step has proved successful, and you can stop safely at any time, and move backwards in the plan at any time. Contrast this to the fast-but-easy plan of taking scheduled downtime to flip everything over to point to B all at once, and crossing your fingers that it all works when you bring the system back online.

It would be great if our upgrade plan had similar properties, so that kubeadm or any random component could crash at any time during the upgrade, and we would never leave the cluster in an unusable state where human intervention would be required to fix it. But, I don't know how feasible that is with the mechanisms we have today, and don't have specific suggestions on how to get there now, so this is just a general grumbling that I hope we can figure out how to do better.

If I have time to offer concrete suggestions for how to do these steps differently, I'll let you know. :-)
```

timothysc commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -195,7 +212,47 @@ func UpgradeControlPlaneComponents(flags *applyFlags, client clientset.Interface
+
+// PerformStaticPodUpgrade performs the upgrade of the control plane components for a static pod hosted cluster
+func PerformStaticPodUpgrade(client clientset.Interface, waiter apiclient.Waiter, internalcfg *kubeadmapi.MasterConfiguration) error {
+	pathManager, err := upgrade.NewKubeStaticPodPathManagerUsingTempDirs(constants.GetStaticPodDirectory())
```
NewKubeStaticPodPathManagerUsingTempDirs <- Does it really need to be this long?

luxas replied:
blocking right now?
Do you have a better suggestion?

timothysc replied:
NewKubeStaticPods  ... PathManagerUsingTempDirs seems unnecessary.

timothysc replied:
For now just leave it.

timothysc commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -195,7 +212,47 @@ func UpgradeControlPlaneComponents(flags *applyFlags, client clientset.Interface
+		return err
+	}
+
+	if err := upgrade.PerformStaticPodControlPlaneUpgrade(waiter, pathManager, internalcfg); err != nil {
```
The name again is weird.. 
upgrade.PerformStaticPodControlPlaneUpgrade

vs. upgrade.StaticPodControlPlane

luxas replied:
renamed to `StaticPodControlPlane`

timothysc commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -98,3 +104,41 @@ func runPreflightChecks(skipPreFlight bool) error {
 }
+
+// getClient gets a real or fake client depending on whether the user is dry-running or not
+func getClient(file string, dryRun bool) (clientset.Interface, error) {
```
How is this in this commit, I've reviewed so much code, I'm positive that this is elsewhere.

timothysc commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -98,3 +104,41 @@ func runPreflightChecks(skipPreFlight bool) error {
+}
+
+// InteractivelyConfirmUpgrade asks the user whether they _really_ want to upgrade.
+func InteractivelyConfirmUpgrade(question string) error {
```
this does not look rebased.

timothysc commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -177,3 +180,17 @@ func GetAdminKubeConfigPath() string {
+// CreateTempDirForKubeadm is a function that creates a temporary directory under /etc/kubernetes/tmp (not using /tmp as that would potentially be dangerous)
+func CreateTempDirForKubeadm(dirName string) (string, error) {
+	// creates target folder if not already exists
+	if err := os.MkdirAll(TempDirForKubeadm, 0700); err != nil {
```
root only on TempDir?

luxas replied:
yes; we concluded that there are possible, pretty nasty `/tmp`-dir attacks where lots of users can do things; so we moved the upgrade temp dir to `/etc/kubernetes/tmp` with root-only access

timothysc reviewed (commented):
Something is wrong, this can't be properly rebased all the way on this commit.  bbdda551b6985a7b5dacf64b95052958d590c822

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -110,15 +119,18 @@ func RunApply(flags *applyFlags) error {
-	// TODO: Implement a prepulling mechanism here
+	// Use a prepuller implementation based on creating DaemonSets
+	// and block until all DaemonSets are ready; then we know for sure that all control plane images are cached locally
+	prepuller := upgrade.NewDaemonSetPrepuller(upgradeVars.client, upgradeVars.waiter, internalcfg)
```
It is pretty crucial for the predictability of this command; there are lots of network failures that might happen, and I want to minimize time that something can go wrong in the actual upgrade process.
For me, this is strictly necessary.

Rationale etc. was in the original proposal as well.

luxas commented:
Removing the do-not-merge label as this is now ready for merge!! :tada:
Only has a small dep on https://github.com/kubernetes/kubernetes/pull/51130, but that doesn't mock much with this PR; can be LGTM'd now if you feel for that.

PTAL @kubernetes/sig-cluster-lifecycle-pr-reviews

timothysc reviewed (approved):
/lgtm due to time.  

We need to get some e2e tests in place and talk with the release team on this.

luxas commented:
fixed some golint failures

luxas committed:
Fully implement the kubeadm upgrade functionality

luxas committed:
Add unit tests for kubeadm upgrade

luxas committed:
autogenerated bazel

luxas commented:
Rebased this now

/retest

kad commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -57,7 +61,196 @@ type ClusterState struct {
+		//                                                                           v1.9.0-alpha.0                                             -> v1.9.0-alpha.1 -> v1.9.0-alpha.2
+
+		// Get and output the current latest unstable version
+		latestVersionStr, latestVersion, err := versionGetterImpl.VersionFromCILabel("latest", "experimental version")
```
Nit: I wouldn't really call "latest" as "experimental". In case of `ci/*` builds I agree to call them experimental, but alphas/betas already have some quality tags attached (even were not very high).

luxas replied:
They are very experimental -- they are just cut from master without any special coordination that feature X, Y, Z must be there for it to ship.

kad commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -57,7 +61,196 @@ type ClusterState struct {
+
+		minorUnstable := latestVersion.Components()[1]
+		// Get and output the current latest unstable version
+		previousBranch := fmt.Sprintf("latest-1.%d", minorUnstable-1)
```
We wouldn't hit it now, but for future (2.x versions), we probably want to have boundary check here.
Worth to add TODO in comments ?

luxas replied:
this is in the queue already so I won't update it anymore.
I think this is probably the smallest problem we'll have regarding 2.x ;)

jamiehannaford commented:
@luxas Did we implement a solution for the discussion outlined in https://github.com/kubernetes/kubernetes/pull/48899#issuecomment-325017058? If not, can we track this as a separate issue for 1.9?

luxas commented:
@jamiehannaford please open a new issue for that
